### DISCUSSION TITLE
Homing using encoder index pulse

### DISCUSSION THREAD

**Discussion by EhsanRasouli1371** [Peter-van-Tol/LiteX-CNC#98] [Q&A]
When I run homing I face the following problem.
By reading the limit switch, the motor moves until receiving the index pulse, and after receiving the encoder index pulse, the axis becomes zero and the motor makes an additional movement.
How can I prevent this extra movement?

Thanks

https://github.com/Peter-van-Tol/LiteX-CNC/assets/147668561/73ef598b-5960-4ba9-9dea-542cb8f4b296

**Comment by EhsanRasouli1371** [comment]
Thank you very much for your reply in the link below
https://github.com/Peter-van-Tol/LiteX-CNC/discussions/35#discussioncomment-9886625
But unfortunately, I did not understand how to solve this problem with your explanation.

**Reply by Peter-van-Tol** [reply]
There is no solution for the overshoot. See the [LinuxCNC manual](https://linuxcnc.org/docs/html/config/ini-homing.html):
> **6.8. HOME**
> The position that the joint will go to upon completion of the homing sequence. After detecting the home switch or home switch then index pulse (depending on configuration), and setting the coordinate of that point to HOME_OFFSET, LinuxCNC makes a move to HOME as the final step of the homing process. The default value is zero. Note that even if this parameter is the same as HOME_OFFSET, the joint will slightly overshoot the latched position as it stops. Therefore there will always be a small move at this time (unless HOME_SEARCH_VEL is zero, and the entire search/latch stage was skipped). This final move will be made at the joint’s maximum velocity unless HOME_FINAL_VEL has been set.

Your design of the machine has to allow for some overshoot. The position after homing is trivial. For example: my machine backs off from the home switches by 5 mm (although I do that with HOME_OFFSET). When the machine is turned off, it returns to the origin (0,0,0). IF this would be the location where the home switches are, the machine would not start up as the homing switches are shared between axes.

**Reply by EhsanRasouli1371** [reply]
Peter, I configured HAL and INI as follows:
HAL:
```
#HOME
net x-index-inable   5A-75B:V8.2.encoder.00.index-enable  
net x-index-inable   joint.0.index-enable  
#net x-index-inable   5A-75B:V8.2.encoder.00.reset
net home-x          <= 5A-75B:V8.2.gpio.in18.in
net home-x           => joint.0.home-sw-in
```
INI:
```
HOME_OFFSET = 0.000000
HOME = 0.000
HOME_SEARCH_VEL = 3
HOME_LATCH_VEL = -1
HOME_FINAL_VEL = 2
HOME_USE_INDEX = TRUE
#HOME_NO_ENCODER_RESET = TRUE
#VOLATILE_HOME = 1
HOME_SEQUENCE = 0
```
And I want the machine to stop immediately after receiving the index pulse and consider that position as zero.
Can this be done?

**Reply by Peter-van-Tol** [reply]
Setting zero after you've stopped gives you a non-repeatable home position. That's not something you would want! As far as I know this is therefore also not possible in LinuxCNC.

What you might consider is to make two configurations:
- start in config 1 and home and then move to position the machine at the origin;
- keep engines enabled, but close LinuxCNC
- (re-)start in config 2 with the settings in velocity method.

When restarting the Litex-CNC FPGA the speed is clamped to 0, so it will be definitely at a stand still and no humming.

**Comment by EhsanRasouli1371** [comment]
Hello guys,  
I set the X-axis as a closed loop and the Z-axis as an open loop.  
When I perform the homing operation, the X-axis works correctly and stops immediately after receiving the encoder index pulse according to the INI file settings.  
However, the homing operation does not work properly on the Z-axis.
It seems that this is because the encoder data is not received in open-loop mode, and this discrepancy between the axis data and the encoder causes an additional movement after receiving the index pulse.  
Is there a solution to zero the axis value immediately upon receiving the index pulse during the homing operation?
[1.txt](https://github.com/user-attachments/files/18595011/1.txt)
[2.txt](https://github.com/user-attachments/files/18595012/2.txt)

**Reply by Peter-van-Tol** [reply]
I think this has been resolved in version 1.3.0 and later (after merging #124). In the HAL, there are two new pins:
- `index-enable`: When true, a rising edge will reset the counter of the stepgen to zero.
- `index_pulse`: When true, a rising edge has been detected on the FPGA. This flag will be active until the index-enable is set to False.

To have this work, you have to add the following lines to your HAL (replacing the current config of the `index-enable` for you Z-axis:
```
net z-index-enable joint.1.index-enable 
net z-index-enable =>  cnc_controller.encoder.02.index-enable 
net z-index-enable =>  cnc_controller.stepgen.02.index-enable 
```

Normally, the index pulse of your encoder will take a round trip from `encoder -> firmware -> LinuxCNC -> firmware -> stepgen`, which is not directly resetting the counter to 0. We have to tie the index-pulse of the encoder to the stepgen in the firmware. This is done by modifying your `.json`-configuration:
``` json
...
"modules": [
    ...,
    {
        "module_type": "stepgen",
        "instance": [
            {
                "pins" : {
                    "stepgen_type": "step_dir",
                    "step_pin": "j9:0",
                    "dir_pin": "j9:1",
                    "index_pin": "j9:2"
                },
                "max_frequency": 400000,
                "soft_stop": true
            },
        ...
        ]
    },
    ...
]
...
```

It should be noted that the `index_pin` of the `stepgen` cannot be the same as the `pinZ` of the corresponding `encoder`. For this soultion you have to splice the wire from the encoder and connect it to two pins on your FPGA. I realize this is sub-optimal, but LiteX does not allow re-use of pins at this moment.
